src/transaction/log_replication_atomic.cpp

```diff
@@ -402,37 +402,72 @@ namespace cublog
 	assert_release (false);
       }
 
-#if !defined (NDEBUG)
-    /* there will be no duplicated object in the map */
-    auto [begin, end] = m_locked_classes.equal_range (trid);
-    if (std::count_if (begin, end, [classoid] (const auto &it)
-    {
-      return it.second == *classoid;
-    }) > 0)
-    {
-      assert (false);
-    }
-#endif
+    m_locked_classes[trid].emplace (*classoid);
+  }
+
+  void
+  atomic_replicator::bookkeep_classname (cubthread::entry &thread_entry, const OID *classoid)
+  {
+    char *classname = NULL;
+
+    assert (!OID_ISNULL (classoid) && !OID_ISTEMP (classoid));
 
-    m_locked_classes.emplace (trid, *classoid);
+    (void) heap_get_class_name (&thread_entry, classoid, &classname);
+    if (classname != NULL)
+      {
+	m_classname_map.emplace (*classoid, classname);
+	free_and_init (classname);
+      }
+    else
+      {
+	/* if classname == NULL, then it is CREATE TABLE/VIEW case */
+      }
+  }
+
+  void
+  atomic_replicator::update_classname_cache_for_ddl (cubthread::entry &thread_entry, const OID *classoid)
+  {
+    /* update locator_Mht_classnames only if needed */
+    char *classname = NULL;
+    (void) heap_get_class_name (&thread_entry, classoid, &classname);
+    if (classname != NULL)
+      {
+	if (m_classname_map.find (*classoid) == m_classname_map.end ())
```

Comment:
Call `find`, store result it in an iterator variable. Reuse below instead of `m_classname_map[*classoid]`.
Every access to the `find`/`operator[]` actually does a search in the hash-map underneath.